### PR TITLE
[PD] Transfer hidden states for mtp when disaggregation

### PR DESCRIPTION
**Pull request by Atream**
## Motivation

When both MTP and PD separation are enabled at the same time, it’s required to pass the hidden states from the prefill node to the decode node as well. The existing implementation currently uses random hidden states, which can lead to errors or a reduction in acceptance rates.

## Modifications

transfer hidden states from Prefill instance to Decode instance.

This PR is based on commit PR https://github.com/sgl-project/sglang/pull/6081.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by Atream** [comment]
This PR is based on commit ff07187d1bcb9fe333a76e5fc0fee4c5a8b00641 from PR #6081.Please merge that PR first.
Currently, there is a lot of hard - coded stuff in the code.When DP Attn is enabled in the decode node, it will hang when a rank becomes idle.I'll address these issues shortly.

**Comment by Atream** [comment]
Reproduce command

```
# prefill 0 change --node-rank
local_ip=$(hostname --ip-address)
MC_TE_METRIC=true SGLANG_TBO_DEBUG=1 python3 -m sglang.launch_server --model-path /mnt/shared-fs/models/deepseek-ai/DeepSeek-V3-0324 --disaggregation-ib-device mlx5_1 --disaggregation-mode prefill --tp-size 8 --dp-size 8 --enable-dp-attention --decode-log-interval 1 --page-size 1 --host ${local_ip} --trust-remote-code --disable-radix-cache --watchdog-timeout 1000000 --mem-fraction-static 0.85 --chunked-prefill-size 262144 --max-running-requests 128 --max-total-tokens 286720 --context-length 71680 --disable-shared-experts-fusion --speculative-algo NEXTN --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 --nnodes 1 --node-rank 0


# decode 0 change --node-rank
local_ip=$(hostname --ip-address)
MC_TE_METRIC=true SGLANG_TBO_DEBUG=1 python3 -m sglang.launch_server --model-path /mnt/shared-fs/models/deepseek-ai/DeepSeek-V3-0324 --disaggregation-ib-device mlx5_1 --disaggregation-mode decode --tp-size 8 --dp-size 8 --enable-dp-attention --decode-log-interval 1 --page-size 1 --host ${local_ip} --trust-remote-code --disable-radix-cache --watchdog-timeout 1000000 --mem-fraction-static 0.85 --chunked-prefill-size 524288 --max-running-requests 128 --max-total-tokens 286720 --context-length 71680 --disable-shared-experts-fusion --speculative-algo NEXTN --speculative-num-steps 3 --speculative-eagle-topk 1 --speculative-num-draft-tokens 4 --nnodes 1 --node-rank 0

python benchmark/gsm8k/bench_sglang.py --port 8000 --parallel 400 --num-questions 400
```

gsm8K score: 93.8
accept length: 2~3 at average

**Comment by Atream** [comment]
dp attention on decode instance is supported now

**Comment by ayrnb** [comment]
:thumbsup::thumbsup::thumbsup:https://github.com/sgl-project/sglang/issues/6684. It seems the root cause of this issue, and I update some code based on your branch. if needed, you can cherry pick it.  [code](https://github.com/bytedance-iaas/sglang/tree/fix/transfer_logit_hidden_state)

**Comment by Atream** [comment]
> 👍👍👍#6684. It seems the root cause of this issue, and I update some code based on your branch. if needed, you can cherry pick it. [code](https://github.com/bytedance-iaas/sglang/tree/fix/transfer_logit_hidden_state)

Thank you for your code. I will check it and make this branch ready to merge asap.

**Comment by ShangmingCai** [comment]
Great job! Will finish the review today. CC: @ByronHsu

**Comment by whybeyoung** [comment]
LGTM

File: python/sglang/srt/disaggregation/utils.py [hunk 1]
```diff
@@ -104,6 +106,11 @@ def __init__(
             # We transfer the metadata of first output token to decode
             # The minimal size for RDMA is 64Bytes, so we pad it to > 64Bytes
             self.output_ids = torch.zeros((size, 16), dtype=torch.int32, device=device)
+
+            # TODO: pad to 64B
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/utils.py]: Since the hidden states will be larger than 64 Bytes, maybe we don't need this todo.

File: python/sglang/srt/disaggregation/prefill.py [hunk 2]
```diff
@@ -402,6 +404,18 @@ def process_batch_result_disagg_prefill(
...
+                    logits_output.hidden_states[
+                        hidden_state_offset : (
+                            hidden_state_offset := hidden_state_offset
+                            + len(req.origin_input_ids)
+                        )
+                    ][-1]
+                    .cpu()
+                    .clone()
+                    if logits_output.hidden_states is not None
+                    else None
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/prefill.py]: Should the slice operation with [-1]: 
```python
                    [
                        hidden_state_offset : (
                            hidden_state_offset := hidden_state_offset
                            + len(req.origin_input_ids)
                        )
                    ][-1]
```
be changed to
```python
last_hidden_index = hidden_state_offset + len(req.origin_input_ids) - 1
        req.hidden_states_tensor = (
            logits_output.hidden_states[last_hidden_index]
            .cpu()
            .clone()
            if logits_output.hidden_states is not None
            else None
        )
        hidden_state_offset += len(req.origin_input_ids)
```

It helps reduce slice copying and speeds up the operation.

File: python/sglang/srt/disaggregation/utils.py [hunk 3]
```diff
@@ -120,20 +127,23 @@ def __init__(
     def get_buf_infos(self):
         ptrs = [
             self.output_ids.data_ptr(),
+            self.output_hidden_states.data_ptr(),
```
Comment (ShangmingCai) [inline python/sglang/srt/disaggregation/utils.py]: Brainstorm: if we set this to None when the spec algorithm is None, maybe we can modify the `send_aux` func to skip sending hidden states for normal PD usage without spec decode. But anyway, we can revisit this later if it really has an impact on the transferring speed. Current impl should be fine.

File: python/sglang/srt/managers/schedule_batch.py [hunk 4]
```diff
@@ -583,6 +583,7 @@ def __init__(
                 self.output_token_ids_logprobs_idx
             ) = None
         self.hidden_states: List[List[float]] = []
+        self.hidden_states_tensor = None  # TODO: use tensor instead of list when needed
```
Comment (ShangmingCai) [inline python/sglang/srt/managers/schedule_batch.py]: Maybe check whether this is a to-do or a note? And we should be specific declaring that we need this "for disagg" instead of "when needed".

**Review by ShangmingCai — COMMENTED** [review]
LGTM. Left some minor comments.

**Review by ShangmingCai — APPROVED** [review]
